endgame_models/get_differences.py

```diff
@@ -0,0 +1,99 @@
+from dataclasses import dataclass
+from typing import Optional, TypeVar
+
+from typing_extensions import Type
+
+from .models import BaseInitialParams
+
+Flat = str | float | int | bool
+
+JSONType = Flat | list["JSONType"] | dict[str, "JSONType"] | tuple["JSONType", ...]
+DictType = dict[str, JSONType]
+ListType = list[JSONType]
+TupleType = tuple[JSONType, ...]
+
+
+@dataclass
+class ReadOnlyDiff:
+    original: JSONType
+    new: JSONType
+
+
+OutputReadOnly = dict[str, "ReadOnlyDiff | OutputReadOnly"]
+
+
+def _output_read_only_diff(
+    initial_dict: DictType, update_dict: DictType, model: Type[BaseInitialParams]
+) -> Optional[OutputReadOnly]:
+    changed_read_only: OutputReadOnly = {}
+    for k, new_v in update_dict.items():
+        if (old_v := initial_dict.get(k)) is not None:
+            assert k in model.__fields__
+            # Update and initial both have parameter
+            field = model.__fields__[k]
```

Comment:
Not a lot different, but maybe it would be better to access fields once: ```field = model.__fields__.get(k); assert field``` ?

Reply:
I would probably do `assert field is not None` (to avoid potential falsey fields, but tbh I can't see much difference:

```py
field = model.__fields__.get(k)
assert field is not None
```
vs
```py
assert k in model.__fields__
field = model.__fields__[k]
```
what's the difference?